### ISSUE TITLE
Hard-negative mining using __getitem__ directive in Dataset class

### ISSUE THREAD

**Issue by yannadani** [pytorch/pytorch#2407]
If I wish to do online hard negative mining (for triplet based algorithms), then I wish to send the recent state dict of my network to the data loader to evaluate the criterion and select the negative. Currently, you need to have a lot of workaround for this to get it working. I think the major limitation is the usage of the __getitem__ directive. Can we have a more flexible approach which allows to send arbitrary number of parameters? Or is there a solution in the current situation which I am not aware of? 

Thanks.

**Comment by lucabergamini** [comment]
Since you are using a generator, you could pass the network instance to the init method and do your evaluations directly in the ```__getitem__``` method, or just update a field of your generator with the dict you need to find the negative margin.

**Comment by soumith** [comment]
@yannadani i think Luca's solution should work, wdyt? You can send the network's state_dict into the data loader's `__init__` and sharing should be preserved.

**Comment by yannadani** [comment]
@soumith yes, right. I get it now. But what exactly do you mean by sharing should be preserved? 
One other issue tangentially related to this post is this: When i try to implement this way, data prefetch and loading is done on CPU, while the network is trained on GPU. Hence, I still seem to have to do a lot of workaround. Do you have any thoughts on this?

**Comment by lucabergamini** [comment]
Well the only workaround you should have is to load on GPU (.cuda method) and unload to the CPU (.cpu method).

**Comment by apaszke** [comment]
@yannadani do you want to run your model in workers on CPU (so that it doesn't interfere with training), or on GPU (so that it's not slow)?
